cEP-0026.md

```diff
@@ -0,0 +1,189 @@
+
+    @classmethod
+    def open(cls, filename, newline=None):
+        return cls(filename, newline=newline)
+
+    def file_as_bytes(self):
+        with open(self.filename, 'rb') as fp:
+            return fp.read()
+
+    def file_as_string(self):
+        return self.file_as_bytes().decode()
+
+    def file_as_list(self):
+        self.lines = linecache.getlines(self.filename)
+        self.linecount = len(self.lines)
+
+    def get_line(self, line):
+        return linecache.getline(self.filename, line)
+
+    @property
+    def timestamp(self):
+        # get the last modified timestamp of the file
+        return os.path.gettime(self.filename)
+
+    def hash(self):
+        # hash the absolute file path to get a unique hash
+        return persistent_hash(self.filename)
+```
+
+### Prototype of Directory class implementation
+
+The objects of this class will also reside in the file-set (formerly known as
+the file-dict) The objects of this class can also be used by `DirectoryBears` as
+discussed on this thread [#4999](https://github.com/coala/coala/issues/4999) by
+performing a simple type check on the entries in the dict, whether the objects
+are instance of `FileFactory` or `Directory`.
+
+Following is the prototype for the `Directory` class.
+
+```python
+import os
+
+from coalib.parsing.Globbing import glob
+
+
+class Directory:
+    def __init__(self, path):
+        self.path = path
+        self.parent = os.path.abspath(os.path.join(p, os.pardir)) # parent dir
+        self.files = 0 # number of files contained in a directory
+        self.subdirs = 0 # number of sub-directories contained in a directory
+
+    def count_children(self):
+        children = os.listdir(self.path)
+
+        for child in children:
+            if os.path.isfile(child):
+                self.files += 1
+
+        self.subdirs = len(children) - self.files
+
+    def get_dir(pattern):
+        return glob(pattern) # get sub directory matching the given glob pattern
+```
+
+### Ignore directories functionality
+
+This will probably be the biggest performance boost for coala. Instead of
+ignoring only unmodified files on subsequent runs, coala shall also ignore
+directories with no unchanged files and subdirectories. This will free up the
+unnecessary space in the cache and speed up runs. It can be implemented
+accordingly:
+
+- `FileFactory` will have an additional attribute parent
+  (the parent directory path) in its `__init__` method.
+- `Directory` class will also have additional attributes `files`
+  (number of files) `subdirs` (number of subdirs) and `parent`
+  (the parent directory's path; in case of project's root it will be `None`)
+- There will be a separate dict (called directory-tracker) containing directory
+  paths as keys and number of files and subdirs as a list of values for e.g.
+  ```
+      { Directory.path: [Directory.files, Directory.subdirs] }
+  ```
+- First all the unchanged files will be deleted from the cache whenever a file
```

Comment:
"all the unchanged files be deleted from the cache"
That sounds like you want to render the cache useless by removing the cache entry when we actually want it, when the file hasn't been changed ;)

Reply:
Ah! you are right. I don't know why I was thinking the other way round. I'll change that.